hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImpl.java

```diff
@@ -337,7 +337,7 @@ void reloadSCMState()
   public void shutdown() throws IOException {
     if (ratisServer != null) {
       ratisServer.stop();
-      ratisServer.getSCMStateMachine().close();
+      ratisServer.getSCMStateMachine().stop();
```

Comment:
>Is the idea behind changing from close() to stop() here that the process should be terminated after all the services are stopped?

yes, i think it is more graceful to stop all the services before SCM process is terminated. after all the services are stopped, we can terminate SCM process explicitly outside "stop" by calling "ExitUtils#terminate". please see this https://github.com/apache/ozone/blob/ebbe9b581542af0cdc0c813996d498678cc78114/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/BackgroundPipelineCreator.java#L141

Reply:
Then shouldn't the terminate be called after all the services are stopped in StorageContainerManager. Wouldn't that be the right place to call terminate? 
_BackgroundPipelineCreator.stop()_ will be called when _PipelineManager_ is closed in _StorageContainerManager_ (line 1523). _SCMMetadataStore_ is stopped after _PipelineManager_ is closed and so it will not be stopped properly.
Please let me know if I am missing something here.

Reply:
thanks for pointing out this , this is indeed not the right place to call `terminate`. i think we can put `terminate` in `scm#stop`, so when `scm#stop` is called , all the service will be stopped and then the SCM process will be terminated

Reply:
We can attempt to stop the services before shutting down. 
But either way, I think the terminate should happen in StorageContainerManager as that is the base service which instantiates and has a handle for all the services. So each sub-service should call back StorageContainerManager to terminate the process.